File: library/ios/src/main/swift/PriceMapping.swift
//
//  PriceMapping.swift
//  DynamicPrice
//
//  Created on 5/21/26.
//  Copyright © 2026 Nimbus Advertising Solutions Inc. All rights reserved.
//

public protocol PriceMapping {
    func getTarget(price: Double) -> String
}

/// A PriceMapping using a linear step function to generate the target value
public struct LinearPriceGranularity {
    public let min: Int
    public let max: Int
    public let step: Int
}

/// A PriceMapping composed of multiple LinearPriceGranularities in sorted order
public struct LinearPriceMapping : PriceMapping {
    public let granularities: [LinearPriceGranularity]

    public init(_ mapping: [LinearPriceGranularity]) {
        self.granularities = mapping.sorted(by: { $0.min < $1.min })

        if granularities.count > 1 {
            let _ = granularities.dropFirst(1).reduce(granularities[0]) {
                guard $0.max == $1.min else {
                    fatalError("Use contiguous granularities: \($0.max) != \($1.min)")
                }
                return $1
            }
        }
    }

    public func getTarget(price: Double) -> String {
        let priceInCents = Int(price * 100)
        let bucket = granularities.first(where: { priceInCents < $0.max }) ?? granularities.last!
        return "\(min(max(priceInCents - priceInCents % bucket.step, bucket.min), bucket.max))"
    }
}

File: library/android/src/androidShared/kotlin/PriceMapping.kt
package com.adsbynimbus.dynamicprice

import kotlin.math.roundToInt
import kotlin.ranges.IntProgression.Companion.fromClosedRange

fun interface PriceMapping {
    fun getTarget(price: Double): String
}

/** A PriceMapping using a linear step function to generate the target value */
@JvmInline
value class LinearPriceGranularity
    private constructor(@PublishedApi internal val value: IntProgression) {

    /**
     * Constructor
     *
     * @param min the minimum line item bid in cents
     * @param max the maximum line item bid in cents
     * @param step the step size
     */
    constructor(min: Int, max: Int, step: Int) : this(fromClosedRange(min, max, step))

    inline val min: Int get() = value.first
    inline val max: Int get() = value.last
    inline val step: Int get() = value.step
}

/** A PriceMapping composed of multiple LinearPriceGranularities in sorted order */
@JvmInline
value class LinearPriceMapping(val granularities: Array<LinearPriceGranularity>) :
    PriceMapping, Comparator<LinearPriceGranularity> {

    init {
        granularities.sortWith(this)
    }

    internal inline val Double.inCents get() = (this * 100).roundToInt()

    override fun getTarget(price: Double): String = with(price.inCents) {
        val bucket = granularities.firstOrNull { this < it.min } ?: granularities.last()
        return "${(this - this % bucket.step).coerceIn(bucket.min, bucket.max)}"
    }

    override fun compare(o1: LinearPriceGranularity, o2: LinearPriceGranularity): Int =
        o1.min - o2.min
}

internal inline val LegacyPriceMapping.toPriceMapping: LinearPriceMapping
    get() = LinearPriceMapping(Array(granularities.size) { i ->
        LinearPriceGranularity(granularities[i].min, granularities[i].max, granularities[i].step)
    })

internal typealias LegacyPriceMapping =  com.adsbynimbus.lineitem.LinearPriceMapping
